cylc/uiserver/workflows_mgr.py

```diff
@@ -99,69 +106,181 @@ def __init__(self, uiserver, context=None):
+                inactive.add(wid)
+                if wid in active_before:
+                    yield (wid, 'active', 'inactive', flow)
+                elif wid in inactive_before:
+                    pass
+                else:
+                    yield (wid, None, 'inactive', flow)
+
+            elif (
+                wid in self.active
+                and flow[CFF.UUID] != self.active[wid].get(CFF.UUID)
+            ):
+                # this flow is running but it's a different run
+                active.add(wid)
+                yield (wid, 'active', 'inactive', flow)
+                yield (wid, 'inactive', 'active', flow)
+
+            else:
+                # this flow is running
+                active.add(wid)
+                if wid in active_before:
+                    pass
+                elif wid in inactive_before:
+                    yield (wid, 'inactive', 'active', flow)
+                else:
+                    yield (wid, None, 'active', flow)
+
+        for wid in active_before - (active | inactive):
+            yield (wid, 'active', None, None)
+
+        for wid in inactive_before - (active | inactive):
+            yield (wid, 'inactive', None, None)
+
+        # return active, inactive
+
+    async def _register(self, wid, flow):
+        """Register a new workflow with the data store."""
+        print(f'_register({wid})')
+        await self.uiserver.data_store_mgr.register_workflow(
+            wid, flow['name'], self.owner
+        )
+
+    async def _connect(self, wid, flow):
+        """Open a connection to a running workflow."""
+        print(f'_connect({wid})')
+        self.active[wid] = flow
+        flow['req_client'] = SuiteRuntimeClient(flow['name'])
+        await self.uiserver.data_store_mgr.sync_workflow(
+            wid,
+            flow['name'],
+            flow[CFF.HOST],
+            flow[CFF.PUBLISH_PORT]
+        )
+
+    async def _disconnect(self, wid):
+        """Disconnect from a running workflow."""
+        print(f'_disconnect({wid})')
+        client = self.active[wid]['req_client']
+        with suppress(IOError):
+            client.stop(stop_loop=False)
+
+    async def _unregister(self, wid):
+        """Unregister a workflow from the data store."""
+        print(f'_unregister({wid})')
+        self.uiserver.data_store_mgr.purge_workflow(wid)
+
+    async def _stop(self, wid):
+        """Mark a workflow as stopped.
+
+        The workflow can't do this itself, because it's not running.
+        """
+        self.uiserver.data_store_mgr.stop_workflow(wid)
+
+    async def update(self):
+        """Scans for workflows, handles any state changes.
+
+        Possible workflow states:
+            active:
+                (i.e. running, held, stopping)
+            inactive:
+                (i.e. stopped or registered)
+            None:
+                (i.e. unregistered)
+
+        Between scans workflows can jump from any state to any other state.
```

Comment:
There are 3 states and 2 scans so that gives us 6 possible state changes.

This has caused me much more of a headache than I should probably admit to so I've explicitly handled each in turn.

Reply:
headache over 👍